kernel: rtl8367, rtl8367b: set the appropriate cpu_port value based on the use of realtek,extif0 to extif2

namiltd opened the pull request:
For rtl8367: Set the appropriate cpu_port value based on the use of realtek,extif0 or realtek,extif1 parameter.
For rtl8367b: Set the appropriate cpu_port value based on the use of realtek,extif0 to realtek,extif2 instead of the additional cpu_port parameter.

Tested on ramips: TP-Link Archer C5 v4
Other routers with the realtek definition,extif0-2 were also checked. For all cpu_port matches that defined in 02_network (last two columns below).
<body>

 realtek,rtl8367 | chip | dts→cpu_port | dts→extif | according to dts→extif (this patch) | 02_network→@&ZeroWidthSpace;eth0
-- | -- | -- | -- | -- | --
ar7242_tplink_tl-wr2543-v1 | RTL8367R | not defined (9) | 0 | **9** | 9
rt3662_asus_rt-n56u | RTL8367M | not defined (9) | 1 | **8** | 8
rt3662_edimax_br-6475nd | RTL8367R | not defined (9) | 0 | **9** | 9
rt3662_samsung_cy-swr1100 | RTL8367R | not defined (9) | 0 | **9** | 9
</body>
<body>

 realtek,rtl8367b | chip | dts→cpu_port | dts→extif | according to dts→extif (this patch) |  02_network→@&ZeroWidthSpace;eth0
-- | -- | -- | -- | -- | --
mt7620a_tplink_archer-c2-v1 | RTL8367RB | 6 | 1 | **6** | 6
mt7620a_tplink_archer-c5-v4 | RTL8367S | 7 | 2 | **7** | 7
mt7620a_tplink_ec220-g5-v2 | RTL8367S | 7 | 2 | **7** | 7
mt7620a_zyxel_keenetic-viva | RTL8367RB | 7 | 2 | **7** | 7
Netgear-wndap620 | RTL8363SB | 5 | 0 | **5** | 5
Netgear-wndap660 | RTL8363SB | 5 | 0 | **5** | 5
mt7620a_iodata_wn-ac733gr3 | RTL8367RB | not defined (5) | 1 | **6** | 6
rt3662_dlink_dir-645 | RTL8367RB | not defined (5) | 1 | **6** | 6
rt3883_belkin_f9k1109v1 | RTL8367R-VB | not defined (5) | 1 | **5** | 5
</body>

namiltd commented:
@hauke Please take a look at this PR.

DragonBluep commented:
Any source code or datasheet can be referenced? Though the result is the same, but it doesn't mean the logic is correct.

namiltd commented:
The result is the same, the logic is simpler, it does not require the definition of cpu_port and at the same time realtek,extif
Just using realtek,extif clearly defines cpu_port, so what's the point of defining it additionally?

namiltd commented on target/linux/generic/files/drivers/net/phy/rtl8367.c:
```diff
@@ -1759,7 +1760,14 @@ static int rtl8367_probe(struct platform_device *pdev)
+	if (of_get_property(smi->parent->of_node, "realtek,extif1", &size))
+		smi->cpu_port = RTL8367_CPU_PORT_NUM - 1;
+	else
+		smi->cpu_port = RTL8367_CPU_PORT_NUM;
```
I will check again the assignments for all routers with realtek,rtl8367 and make a list.

namiltd commented:
<html>
<head>
	
	
	<title></title>
	<meta name="generator" content="LibreOffice 24.2.0.3 (Windows)"/>
	<style type="text/css">
		body,div,table,thead,tbody,tfoot,tr,th,td,p { font-family:"Liberation Sans"; font-size:x-small }
		a.comment-indicator:hover + comment { background:#ffd; position:absolute; display:block; border:1px solid black; padding:0.5em;  } 
		a.comment-indicator { background:red; display:inline-block; border:1px solid black; width:0.5em; height:0.5em;  } 
		comment { display:none;  } 
	</style>
	
</head>

<body>

chipset | router | dts→cpu_port | dts→extif | according to dts→extif | 02_network
-- | -- | -- | -- | -- | --
rtl8367 | ar7242_tplink_tl-wr2543-v1 | not defined (9) | 0 | 9 | 9
rt3662_asus_rt-n56u | not defined (9) | 1 | 8 | 8
rt3662_edimax_br-6475nd | not defined (9) | 0 | 9 | 9
rt3662_samsung_cy-swr1100 | not defined (9) | 0 | 9 | 9
rtl8367b | mt7620a_tplink_archer-c2-v1 | 6 | 1 | 6 | 6
mt7620a_tplink_archer-c5-v4 | 7 | 2 | 7 | 7
mt7620a_tplink_ec220-g5-v2 | 7 | 2 | 7 | 7
mt7620a_zyxel_keenetic-viva | 7 | 2 | 7 | 7
Netgear-wndap620 | 5 | 0 | 5 | 5
Netgear-wndap660 | 5 | 0 | 5 | 5
mt7620a_iodata_wn-ac733gr3 | not defined (5) | 1 | 6 | 6
rt3662_dlink_dir-645 | not defined (5) | 1 | 6 | 6
rt3883_belkin_f9k1109v1 | not defined (5) | 1 | 6 | 5


</body>

</html>

namiltd commented:
<body>

router with rtl8367 | dts→cpu_port | dts→extif | according to dts→extif | 02_network
-- | -- | -- | -- | --
ar7242_tplink_tl-wr2543-v1 | not defined (9) | 0 | 9 | 9
rt3662_asus_rt-n56u | not defined (9) | 1 | 8 | 8
rt3662_edimax_br-6475nd | not defined (9) | 0 | 9 | 9
rt3662_samsung_cy-swr1100 | not defined (9) | 0 | 9 | 9


</body>

namiltd commented:
<body>

 realtek,rtl8367 | chip | dts→cpu_port | dts→extif | according to dts→extif | 02_network→@&ZeroWidthSpace;eth0
-- | -- | -- | -- | -- | --
ar7242_tplink_tl-wr2543-v1 | RTL8367R | not defined (9) | 0 | **9** | 9
rt3662_asus_rt-n56u | RTL8367M | not defined (9) | 1 | **8** | 8
rt3662_edimax_br-6475nd | RTL8367R | not defined (9) | 0 | **9** | 9
rt3662_samsung_cy-swr1100 | RTL8367R | not defined (9) | 0 | **9** | 9

</body>
<body>

 realtek,rtl8367b | chip | dts→cpu_port | dts→extif | according to dts→extif |  02_network→@&ZeroWidthSpace;eth0
-- | -- | -- | -- | -- | --
mt7620a_tplink_archer-c2-v1 | RTL8367RB | 6 | 1 | **6** | 6
mt7620a_tplink_archer-c5-v4 | RTL8367S | 7 | 2 | **7** | 7
mt7620a_tplink_ec220-g5-v2 | RTL8367S | 7 | 2 | **7** | 7
mt7620a_zyxel_keenetic-viva | RTL8367RB | 7 | 2 | **7** | 7
Netgear-wndap620 | RTL8363SB | 5 | 0 | **5** | 5
Netgear-wndap660 | RTL8363SB | 5 | 0 | **5** | 5
mt7620a_iodata_wn-ac733gr3 | RTL8367RB | not defined (5) | 1 | **6** | 6
rt3662_dlink_dir-645 | RTL8367RB | not defined (5) | 1 | **6** | 6
rt3883_belkin_f9k1109v1 | RTL8367R-VB | not defined (5) | 1 | **5** | 5


</body>
In the current version, for all routers on the list, cpu_port from detection is equal to the value read from 02_network file.

DragonBluep commented on target/linux/generic/files/drivers/net/phy/rtl8367.c:
```diff
@@ -1722,6 +1723,13 @@ static int rtl8367_detect(struct rtl8366_smi *smi)
+	if (of_get_property(smi->parent->of_node, "realtek,extif1", &size))
+		smi->cpu_port = RTL8367_CPU_PORT_NUM - 1;
+	else
+		smi->cpu_port = RTL8367_CPU_PORT_NUM;
```
```
	if (of_get_property(smi->parent->of_node, "realtek,extif1", &size))
		smi->cpu_port -= 1;
```

namiltd replied:
At the time of calling probe(), we do not know the chip version yet.

DragonBluep commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -1583,6 +1586,18 @@ static int rtl8367b_detect(struct rtl8366_smi *smi)
+		else
+			smi->cpu_port = RTL8367B_CPU_PORT_NUM + 1;
+	} else
+		smi->cpu_port = RTL8367B_CPU_PORT_NUM;
```
```
		if (of_get_property(smi->parent->of_node, "realtek,extif1", &size) && !rvb)
			smi->cpu_port += 1;
		else if (of_get_property(smi->parent->of_node, "realtek,extif2", &size))
			smi->cpu_port += 2;
```

namiltd replied:
But calling rtl8367_detect twice (or more) would change the value.

DragonBluep commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -1583,6 +1586,18 @@ static int rtl8367b_detect(struct rtl8366_smi *smi)
+	if (of_get_property(smi->parent->of_node, "realtek,extif2", &size))
+		smi->cpu_port = RTL8367B_CPU_PORT_NUM + 2;
+	else if (of_get_property(smi->parent->of_node, "realtek,extif1", &size)) {
+		if (rvb) /* for the RTL8367R-VB chip, extif1 corresponds to cpu_port 5 */ 
```
Maybe directly check `chip_ver` or `chip_name`?

namiltd replied:
What's the point if it's checked a few lines up? It's simpler this way.

DragonBluep reviewed (commented):
I suggest simplifying the if else statement. And perhaps these checks can be moved to `probe()` function?

namiltd commented:
@hauke @PolynomialDivision @robimarko Is there any chance of this PR being accepted?

robimarko commented:
I dont have any knowledge about this IC nor boards that use so I cannot really comment

robimarko commented:
Sorry, closed by accident

namiltd commented:
I understand, thank you for your immediate response.

DragonBluep reviewed (commented):
I am unable to make further comments because I’m not familiar with realtek chips. In my opinion, at least the code logic should be correct.

namiltd committed:
kernel: rtl8367: set the correct cpu_port value

Set the appropriate cpu_port value based on the use of realtek,extif0 or realtek,extif1 parameter.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/15033
Signed-off-by: Robert Marko <robimarko@gmail.com>

namiltd commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -1621,9 +1634,7 @@ static int  rtl8367b_probe(struct platform_device *pdev)
-	if (of_property_read_u32(pdev->dev.of_node, "cpu_port", &smi->cpu_port)
-	    || smi->cpu_port >= smi->num_ports)
-		smi->cpu_port = RTL8367B_CPU_PORT_NUM;
+	smi->cpu_port = RTL8367B_CPU_PORT_NUM;
```
Corrected. I removed duplicate assignments.

namiltd commented on target/linux/generic/files/drivers/net/phy/rtl8367.c:
```diff
@@ -1722,6 +1723,13 @@ static int rtl8367_detect(struct rtl8366_smi *smi)
 	dev_info(smi->parent, "RTL%s ver. %u chip found\n",
 		 chip_name, rtl_ver & RTL8367_RTL_VER_MASK);
 
+	if (of_get_property(smi->parent->of_node, "realtek,extif1", &size))
```
Done.

namiltd commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -1583,6 +1584,18 @@ static int rtl8367b_detect(struct rtl8366_smi *smi)
 
+	if (of_get_property(smi->parent->of_node, "realtek,extif2", &size))
+		smi->cpu_port = RTL8367B_CPU_PORT_NUM + 2;
+	else if (of_get_property(smi->parent->of_node, "realtek,extif1", &size)) {
```
Done.

namiltd committed:
kernel: rtl8367b: remove unnecessary cpu_port setting

Set the appropriate cpu_port value based on the use of realtek,extif0 to extif2
instead of the additional cpu_port parameter.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/15033
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
Thanks! Rebased on top of main and merged!
